Upgrade everything we can for a chain reset

jflatow opened the pull request:
This branch is meant to be an integration point for all changes going into the next release, which at this point is looking to be a brand new testnet chain. We lost our ability to upgrade via governance with the re-org breakage, and hard forking has proved difficult, so we need to spend more time on it (but don't want to block people from using the testnet in the meantime).

At the moment, this will still need to be rebased onto develop, which has some other changes to scenarios which may conflict. There are also a few other branches that should get rebased / merged into this one, and then probably a few fixes to make sure CI still works after everything else is merged.

jflatow committed:
Re-upgrade Substrate

WIP: CASH benchmarks aren't working for some reason (commented out)
 and test is failing due to another account minimal existence issue

Also had to delete GrandpaRpcHandler, but hopefully not needed now?

jflatow committed:
Re-add Gateway addresses, which validators now have

Validators are now required to have a minimum CASH balance in order to set session keys, and therefore validate.
On committing a pipeline change, any account with a balance above the minimum is marked as existing in the Substrate account store.
Any account with a balance below the minumum is purged from the account store.

More test coverage needed.

jflatow committed:
Really use min balance path in tests

jflatow committed:
Attempt to patch issue blocking ingestion

Only filter the blocks that were supported (not dissented), and only allow support or dissent for a given signer.
The vote is being cast only on the support side, dissent is only derived if the chain receives a disagreement with the first one it received.

hayesgm committed:
Test m10 patch

jflatow committed:
Cargo audit fixes

Also upgrade substrate again in order to bump deps. This is using another substrate branch so as not to break the one others are using.

Couldn't update cranelift-codegen without starting to break stuff in substrate.

hayesgm committed:
Start to build code substitutes

jflatow committed:
Fix first build from integration

jflatow commented on runtime/src/lib.rs:
```diff
@@ -119,7 +120,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     impl_name: create_runtime_str!("gateway"),
     authoring_version: 1,
-    spec_version: 10,
+    spec_version: 11,
```
We still want this?

hayesgm replied:
I think this is right.

jflatow commented on integration/__tests__/upgrade_to_m10__hardfork_scen.js:
```diff
@@ -0,0 +1,185 @@
+const { buildScenarios } = require('../util/scenario');
```
Should we just delete this whole scenario or what?

hayesgm replied:
I'd just mark it as skip for now and then delete it separately. Just want to make sure we have something to revive in the future.

hayesgm commented on integration/__tests__/lock_scen.js:
```diff
@@ -23,6 +23,16 @@ let now = Date.now();
 buildScenarios('Lock Scenarios', lock_scen_info, [
   {
     name: 'Lock Collateral',
+    info: {
```
I don't think we need this; just wanted to test it here for re-org testing.

hayesgm commented on integration/babel.config.json:
```diff
@@ -0,0 +1,13 @@
+{
```
I'd like to remove this if I can, but we're stuck with it for now.

hayesgm commented on integration/util/scenario/ctx.js:
```diff
@@ -68,10 +68,13 @@ class Ctx {
-    console.error(msg);
     this.error(msg);
-    process.exit(1);
+    throw(msg);
```
Not sure why we changed this (or why we had it). I had issues with the tests not exiting when a node failed, but maybe this is right now?

jflatow replied:
Was it doing `throw` before, or you think `throw` is actually broken? We were logging like 3x so I replaced 1 of them and the process exit with a `throw` since it seemed cleaner

hayesgm commented on integration/util/scenario/token.js:
```diff
@@ -334,7 +339,7 @@ async function buildTokens(tokensInfoHash, scenInfo, ctx) {
   let etherToken = new EtherToken(scenInfo.eth_liquidity_factor, ctx);
   tokens.push(etherToken);
-  if (scenInfo.eth_supply_cap) {
+  if (!!scenInfo.eth_supply_cap) {
```
I think this is identical to the original, so probably can remove this.

hayesgm commented on integration/util/scenario/validator.js:
```diff
@@ -141,63 +162,80 @@ class Validator {
+    }
+  }
+
+  // !!!
```
Not sure this comment.

hayesgm commented on integration/util/scenario/validator.js:
```diff
@@ -141,63 +162,80 @@ class Validator {
-    }
+    let executionArgs = [
+      '--execution', this.native ? 'Native' : 'Wasm',
+      /*'--wasm-runtime-overrides', this.version.wasmDir()*/
```
We might want to add this back in? Maybe as an option, at the least?

hayesgm commented on pallets/cash/src/chains.rs:
```diff
@@ -561,9 +550,24 @@ impl ChainBlockTally {
+    pub fn has_supporter(&self, validator_id: &ValidatorIdentity) -> bool {
         // note that these set types are not optimized and inefficient
-        self.support.contains(&validator_id) || self.dissent.contains(&validator_id)
+        log!(
```
We can probably remove this. It was for debugging.

hayesgm commented on pallets/cash/src/internal/events.rs:
```diff
@@ -97,6 +97,7 @@ pub fn track_chain_events<T: Config>() -> Result<(), Reason> {
 
 /// Perform the next step of tracking events from an underlying chain.
 pub fn track_chain_events_on<T: Config>(chain_id: ChainId) -> Result<(), Reason> {
+    log!("track_chain_events_on'ing");
```
can probably remove this log, as well.

hayesgm commented on pallets/cash/src/internal/initialize.rs:
```diff
@@ -115,7 +116,7 @@ pub fn initialize_block<T: Config>(now: Timestamp) -> Result<(), Reason> {
 
     <Module<T>>::deposit_event(Event::TransferCash(
-        ChainAccount::Reserved,
+        GATEWAY_VOID,
```
Not sure what this is?

jflatow replied:
This is where we transfer from to miners...and now we have `ChainAccount::Gate(...)` instead of `ChainAccount::Reserved`, so its just shorthand for `ChainAccount::Gate([0u8; 32])`

hayesgm commented on scripts/build_release.sh:
```diff
@@ -3,6 +3,7 @@
 set -eo pipefail
 
 version="$1"
+machine=`node -e 'console.log(process.platform + "-" + process.arch)'`
```
That'll do it.

hayesgm reviewed (approved):
Awesome stuff.

jflatow committed:
New and improved RPCs (#354)

Co-authored-by: Coburn Berry <coburn@compound.finance>

maxwolff committed:
Reorg fixes (#349)

jflatow committed:
Add two basic parser fuzzers and Fuzzbuzz integration (#320)

* Add two basic parser fuzzers and Fuzzbuzz integration

* Upgrade substrate / fix compilation

Co-authored-by: Everest Munro-Zeisberger <evmunro1@gmail.com>

waynenilsen committed:
Update integration testing readme (#333)

hayesgm committed:
Fix lock scen

hayesgm committed:
Build Ethereum Contract to test Trx Request from Chain (#282)

This patch adds a new contract (and a `TrxRequestBuilder`) which tests executing a trx request from on-chain. The builder is rudimentary, but can be the start of a full-time builder. The contract `Locker` just locks Eth and transfers it to you (the sender)'s account on Gateway. I believe this is currenty failing due to `(Transfer Cash Max 0x...)` failing right now.

Note: the one bug discovered so far is that events from a trx are processed in reverse order. We should ensure we fix that in our new event system. (It was)

jflatow committed:
Bump rust version in CI and simplify scenario workflow

hayesgm committed:
Use saddle 0.1.23

coburncoburn committed:
test dissenting (#356)

hayesgm committed:
Fix some tests

hayesgm committed:
Mark failing tests for follow-up

jflatow committed:
Fix fetch_eth_blocks and formulate reorg issue (#357)

* Fix fetch_eth_blocks and formulate reorg issue

We don't need workers to abstain, since that creates different behaviors depending on where the 'invalid blockchain' is found.

Also, avoid fetching duplicates in formulating the reorg.
Also, really only fetch `slack` blocks, not `slack + 1`.

* Fix dissent and concordance scenario

jflatow committed:
Cleanups (#359)

hayesgm committed:
Fix up scenarios for moon branch

* Temp work on scens

* Start to diagnose all remaining scen issues

* Make the re-org test consistent, if slightly less cool

waynenilsen committed:
Fix TOB #2 (#362)
